src/main/windows/mainWindow.ts

```diff
@@ -3,146 +3,236 @@
 import {localizeMessage} from 'main/i18nManager';
 
 import * as Validator from '../Validator';
 import ContextMenu from '../contextMenu';
-import {getLocalPreload, getLocalURLString} from '../utils';
-
-function saveWindowState(file: string, window: BrowserWindow) {
-    const windowState: SavedWindowState = {
-        ...window.getBounds(),
-        maximized: window.isMaximized(),
-        fullscreen: window.isFullScreen(),
-    };
-    try {
-        fs.writeFileSync(file, JSON.stringify(windowState));
-    } catch (e) {
-    // [Linux] error happens only when the window state is changed before the config dir is created.
-        log.error(e);
+import {getLocalPreload, getLocalURLString, isInsideRectangle} from '../utils';
+
+export class MainWindow {
+    private win?: BrowserWindow;
+
+    private savedWindowState: SavedWindowState;
+    private ready: boolean;
+
+    constructor() {
+        // Create the browser window.
+        this.ready = false;
+        this.savedWindowState = this.getSavedWindowState();
+
+        ipcMain.handle(GET_FULL_SCREEN_STATUS, () => this.win?.isFullScreen());
     }
-}
 
-function isInsideRectangle(container: Electron.Rectangle, rect: Electron.Rectangle) {
-    return container.x <= rect.x && container.y <= rect.y && container.width >= rect.width && container.height >= rect.height;
-}
+    init = () => {
+        const windowOptions: BrowserWindowConstructorOptions = Object.assign({}, this.savedWindowState, {
+            title: app.name,
+            fullscreenable: true,
+            show: false, // don't start the window until it is ready and only if it isn't hidden
+            paintWhenInitiallyHidden: true, // we want it to start painting to get info from the webapp
+            minWidth: MINIMUM_WINDOW_WIDTH,
+            minHeight: MINIMUM_WINDOW_HEIGHT,
+            frame: !this.isFramelessWindow(),
+            fullscreen: this.shouldStartFullScreen(),
+            titleBarStyle: 'hidden' as const,
+            trafficLightPosition: {x: 12, y: 12},
+            backgroundColor: '#fff', // prevents blurry text: https://electronjs.org/docs/faq#the-font-looks-blurry-what-is-this-and-what-can-i-do
+            webPreferences: {
+                disableBlinkFeatures: 'Auxclick',
+                preload: getLocalPreload('desktopAPI.js'),
+                spellcheck: typeof Config.useSpellChecker === 'undefined' ? true : Config.useSpellChecker,
+            },
+        });
 
-function isFramelessWindow() {
-    return os.platform() === 'darwin' || (os.platform() === 'win32' && Utils.isVersionGreaterThanOrEqualTo(os.release(), '6.2'));
-}
+        if (process.platform === 'linux') {
+            windowOptions.icon = path.join(path.resolve(app.getAppPath(), 'assets'), 'linux', 'app_icon.png');
+        }
+
+        this.win = new BrowserWindow(windowOptions);
+        this.win.setMenuBarVisibility(false);
+
+        if (!this.win) {
+            log.error('unable to create main window');
```

Comment:
I don't believe so, this should only fail if for some reason the app can't create a window. I'm not even sure where that case happens, maybe it's permissions?

However, you make a good point. Currently all we do is log the error and silently quit, but this should actually throw an error so that the user knows what happened.